src/operator/contrib/sync_batch_norm-inl.h

```diff
@@ -271,69 +273,102 @@ class SyncBatchNorm : public Operator {
-      var = scale * sumall_except_dim<1>(F<mshadow_op::square>(data));
-      SharedND<mshadow::Tensor<cpu, 1, real_t>> *sharedMean =
-        global_shared_mean.Register(param_.key, param_.ndev);
-      SharedND<mshadow::Tensor<cpu, 1, real_t>> *sharedVar =
-        global_shared_var.Register(param_.key, param_.ndev);
-      // copy to cpu, push and pull
-      Tensor<cpu, 1, real_t>* mean_cpu_ptr = sharedMean->Retrieve(mean.shape_, myRank);
-      Tensor<cpu, 1, real_t>* var_cpu_ptr = sharedVar->Retrieve(mean.shape_, myRank);
-      mshadow::Copy(*mean_cpu_ptr, mean, s);
-      mshadow::Copy(*var_cpu_ptr, var, s);
-      sharedMean->SetReady(myRank);
-      sharedVar->SetReady(myRank);
-      global_barrier->Wait();
-      Tensor<cpu, 1, real_t> mean_cpu = sharedMean->Pop(myRank);
-      Tensor<cpu, 1, real_t> var_cpu = sharedVar->Pop(myRank);
-      // copy back to gpu
-      mshadow::Copy(mean, mean_cpu, s);
-      mshadow::Copy(var, var_cpu, s);
-
-      var = var-F<mshadow_op::square>(mean);
-      Assign(out, req[syncbatchnorm::kOut], broadcast<1>(slope, out.shape_) *
-             (data - broadcast<1>(mean, data.shape_)) /
-             F<mshadow_op::square_root>(broadcast<1>(var + param_.eps, data.shape_)) +
-             broadcast<1>(bias, out.shape_));
-    } else {
-      Assign(out, req[syncbatchnorm::kOut], broadcast<1>(slope /
-                                          F<mshadow_op::square_root>(moving_var + param_.eps),
-                                          data.shape_) * data +
-             broadcast<1>(bias - (slope * moving_mean) /
-                          F<mshadow_op::square_root>(moving_var + param_.eps), data.shape_));
-    }
+    MSHADOW_TYPE_SWITCH(in_data[syncbatchnorm::kData].type_flag_, DType, {
+      const bool is_double = std::is_same<DType, double>::value;
+      CHECK_EQ(is_double, false)
+        << "Synchronized BatchNorm does not support double-precision floating number yet...";
+      const real_t scale = static_cast<real_t>(in_data[syncbatchnorm::kData].shape_[1]) /
+        static_cast<real_t>(in_data[syncbatchnorm::kData].shape_.Size());
+      const size_t data_size = in_data[syncbatchnorm::kData].Size();
+      Tensor<xpu, 4> data;
+      Tensor<xpu, 4> out;
+      Tensor<xpu, 1> workspace;
+      if (!std::is_same<DType, real_t>::value) {
+        workspace = ctx.requested[syncbatchnorm::kTempSpace].get_space<xpu, 1>(
+          Shape1(data_size * 2), s);
+      }
+      if (in_data[syncbatchnorm::kData].ndim() == 2) {
+        Shape<4> dshape = Shape4(in_data[syncbatchnorm::kData].shape_[0],
+                                in_data[syncbatchnorm::kData].shape_[1], 1, 1);
+        if (std::is_same<DType, real_t>::value) {
+          data = in_data[syncbatchnorm::kData].get_with_shape<xpu, 4, real_t>(dshape, s);
+          out = out_data[syncbatchnorm::kOut].get_with_shape<xpu, 4, real_t>(dshape, s);
+        } else {
+          data = Tensor<xpu, 4>(workspace.dptr_, dshape, s);
```

Comment:
I think we should write a function such as getData() to handle the different ways to get data between half_t and real_t case. So many repetition if branches in this code makes it hard to read and bug prone.

Reply:
If you think of it carefully you'll see that such a function will have quite a few arguments and still very hard to debug and will introduce extra overhead.

Reply:
I think there are only a few arguments: (TBlob &, DType, Stream &). What is the extra overhead do you mean? On the other hand, I can see this same logic block being used at least 4 times in this code.

Reply:
no you actually need more than that including `in_data[syncbatchnorm::kData]`, `out_data[syncbatchnorm::kOut]`, `s`, `workspace.dptr_`, `dshape` and some of them need to be null/nothing in some cases, so I don't it's a very suitable case for a "common" helper function.

Reply:
Both in_data[syncbatchnorm::kData] and out_data[syncbatchnorm::kOut] are TBlob, you only need to write a common function to process both of them so only one TBlob argument is needed. As for workspace, I think you can wrap that inside the function as well with some proper refactoring.